compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/ExpressionCodegen.kt

```diff
@@ -455,12 +457,49 @@ class ExpressionCodegen(
     }
 
     override fun visitSetField(expression: IrSetField, data: BlockInfo): StackValue {
+        val expressionValue = expression.value
+        if (irFunction is IrConstructor && irFunction.isPrimary) {
+            // Do not add redundant field initializers that initialize to default values.
+            // "expression.origin == null" means that the field is initialized when it is declared,
+            // i.e., not in an initializer block or constructor body.
+            if (expression.origin == null && expressionValue is IrConst<*> && skipDefaultValue(
+                    expression.symbol.owner.type, expressionValue
+                )
+            ) return none()
+        }
         expression.markLineNumber(startOffset = true)
         val fieldValue = generateFieldValue(expression, data)
-        fieldValue.store(expression.value.accept(this, data), mv)
+        fieldValue.store(expressionValue.accept(this, data), mv)
         return none()
     }
 
+    private fun skipDefaultValue(fieldType: IrType, constExpression: IrConst<*>): Boolean {
+        val value = constExpression.value
+        val type = constExpression.asmType
+        if (isPrimitive(type)) {
+            if (!fieldType.isMarkedNullable() && value is Number) {
+                if (type in setOf(Type.INT_TYPE, Type.BYTE_TYPE, Type.LONG_TYPE, Type.SHORT_TYPE) && value.toLong() == 0L) {
+                    return true
+                }
+                if (type === Type.DOUBLE_TYPE && value == 0.0) {
```

Comment:
This is not correct value  == 0.0 check performed via IEEE754,
so 0.0 == - 0.0 (true),  
but 0.0.equals (-0.0 ) (false)

Reply:
Please also add test for such cases

Reply:
Actually this works out fine because `value` is of type `Number` and therefore `value == 0.0` invokes `Number.equals`. But to be explicit, changed to `value.toDouble().equals(0.0)` and added tests.

Tangentially related: When I copy pasted code from https://github.com/JetBrains/kotlin/blob/8eeb9d4faab68b4821bbad62ae38a59b03f16930/compiler/backend/src/org/jetbrains/kotlin/codegen/MemberCodegen.java#L569, I had the Kotlin plugin convert the code as a starting point, and it used `==` instead of `.equals`. Bug in the converter (at least for doubles and floats)? So the simplest repro of this bug is something like this:

```
        Double posZero = 0.0;
        Double negZero = -0.0;
        if (posZero.equals(negZero)) System.out.println("EQUAL");
        else System.out.println("NOT EQUAL");
```

gets converted to
```
        val posZero = 0.0
        val negZero = -0.0
        if (posZero == negZero)
            println("EQUAL")
        else
            println("NOT EQUAL")
```

...which aren't the same. And the editor gives a warning that `if (posZero == negZero)` is always false, which isn't correct either.

Reply:
@punzki Thanks, I've filed KT-29691, KT-29692